### ISSUE TITLE
Unsoundness problem in Node::tree

### ISSUE THREAD

**Issue by lwz23** [DioxusLabs/blitz#169]
Hello, thank you for your contribution in this project, I am scanning the unsoundness problem in rust project.
I notice the following code:
```
pub struct Node {
    // The actual tree we belong to. This is unsafe!!
    pub tree: *mut Slab<Node>,

    /// Our Id
    pub id: usize,
    /// Our parent's ID
    pub parent: Option<usize>,
................................
}
impl Node {
    pub fn tree(&self) -> &Slab<Node> {
        unsafe { &*self.tree }
    }
...........................
}
```
Considering that `tree` is a pub field, I assume that users can directly manipulate this field. This potential situation could result in `self.tree` being a null pointer, and directly dereferencing it might trigger undefined behavior (UB).
PoC:
```
extern crate blitz_dom;
extern crate style;

use blitz_dom::node::Node;
use std::ptr;
use style::{
    data::ElementData,
    properties::{parse_style_attribute, PropertyDeclarationBlock},
    servo_arc::Arc as ServoArc,
    shared_lock::{Locked, SharedRwLock},
    stylesheets::CssRuleType,
};
use blitz_dom::NodeData;
fn main() {
    let node = Node::new(ptr::null_mut(),0,SharedRwLock::new(),NodeData::Document);
    let _=node.tree();
}
```
If there is no external using for `Node`, maybe it should not be marked as `pub`, at least for its field should not be mark as `pub`.

**Comment by nicoburns** [comment]
Good point. I'll make it private. I think it was initially private, but was made public to support a cross-crate access that is no longer necessary.

**Comment by lwz23** [comment]
Hello again, I think maybe current fix is not sufficient :(
because I notice the following function:
https://github.com/DioxusLabs/blitz/blob/017f8ea722df16411a8d53fd340f32cd26bbd585/packages/blitz-dom/src/node.rs#L97
the user can still pass a Null pointer through the `new` method
maybe we should mark new method as private or add check to make sure the pointer is not null?

**Comment by nicoburns** [comment]
I've made the `new` function private in https://github.com/DioxusLabs/blitz/commit/e12a3c25ece6a5f43b762e4ea9e3d263c494b977. But I'll leave this open for now in case there are other issues. That tree pointer is definitely a source of unsafety I'd like to eliminate at some point. But Stylo's design makes that tricky.

**Comment by lwz23** [comment]
Thanks, I will keep search if there are other cases. If I found, I will report it in this issue.

**Comment by lwz23** [comment]
Also, given that this is an Unsound issue, I'm not sure if we should report it to RustSec?

**Comment by nicoburns** [comment]
> Also, given that this is an Unsound issue, I'm not sure if we should report it to RustSec?

I think we should probably avoid that for now. Blitz doesn't have any real users yet, so they won't have run into it. And our wrapper libraries were not triggering the unsafety.

**Comment by lwz23** [comment]
Ok, I'm not quite familiar with the requirements of the report, thanks for clarification.
